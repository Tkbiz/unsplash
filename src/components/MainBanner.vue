<template>
  <div
      v-if="randomPhoto"
      class="container--fluid main-banner">
    <v-row class="justify-end align-center position-relative ma-0">
      <v-container class="banner-block">
        <v-row>
          <v-col
              class="pa-xl-0"
              cols="12"
              md="8"
              lg="5"
              xl="4">
            <h1 v-if="randomPhoto.user"
                class="banner-title col-12 col-lg-10">
              {{ bannerTitle }}
              <br>
              {{ randomPhoto.user.name | truncate(65, '...') }}
            </h1>
            <p v-if="randomPhoto.description"
               class="banner-description col-12 col-md-7 col-lg-8">
              {{ randomPhoto.description | truncate(130, '...') }}
            </p>
            <v-btn
                class="banner-btn"
                depressed>
              explore all
            </v-btn>
          </v-col>
        </v-row>
      </v-container>
      <v-col
          cols="12"
          lg="7"
          class="position-relative banner-img">
        <v-img
            v-if="randomPhoto.urls"
            :src="randomPhoto.urls.regular ? randomPhoto.urls.regular : '@/assets/images/no-image.png'">
        </v-img>
      </v-col>
      <SearchNavigation :tags="tags"/>
    </v-row>
  </div>
</template>

<script>
import SearchNavigation from '@/components/SearchNavigation'
import {truncateTextMixin} from "@/mixins/truncateTextMixin"

export default {
  name: 'MainBanner',
  mixins: [truncateTextMixin],
  props: {
    tags: {
      type: Array,
      required: true
    }
  },
  data: () => ({
    bannerTitle: 'Photo of the Day by ',
    randomPhoto: {}
  }),
  components: {
    SearchNavigation
  },
  async mounted() {
   this.randomPhoto = await this.$store.dispatch('fetchRandomPhoto')
  }
}
</script>
