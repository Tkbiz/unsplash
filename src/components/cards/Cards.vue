<template>
  <v-container class="px-xl-0">
    <v-row class="mx-xl-n4">
      <v-col
          class="d-flex px-xl-4 py-0 mb-8"
          cols="12"
          sm="6"
          md="3"
          lg="3">
        <v-select
            class="custom-select"
            v-model="selectOrientation"
            @change="searchByOrientation"
            :items="items"
            label="Any orientation"
            outlined>
        </v-select>
      </v-col>
    </v-row>
    <v-row class="mx-xl-n4">
      <v-col
          cols="12"
          v-for="(card, index) of photosList"
          :key="card.id + index"
          class="d-flex px-xl-4 py-0 mb-8"
          :class="card.height < 4000 ? 'col-sm-6 col-lg-3': 'col-sm-12 col-lg-6'">
        <CardSmall
            v-if="card.height < 4000"
            :card="card"
            :userName="userName"/>
        <CardLarge
            v-else
            :card="card"
            :userName="userName"/>
      </v-col>
      <Loader v-if="loading"/>
      <Observer
          @intersect="showNewPhotos"
          :options="{root, threshold}"/>
    </v-row>
  </v-container>
</template>

<script>
import CardSmall from '@/components/cards/CardSmall'
import CardLarge from '@/components/cards/CardLarge'
import Observer from '@/components/Observer'
import {bus} from '@/bus'

export default {
  name: 'Cards',
  components: {
    CardSmall,
    CardLarge,
    Observer
  },
  props: {
    photosList: {
      type: Array,
      required: true
    },
    userName: {
      required: true
    }
  },
  data: () => ({
    page: 1,
    newPhotos: [],
    root: null,
    threshold: 0.5,
    loading: false,
    items: ['landscape', 'portrait', 'squarish'],
    selectOrientation: null
  }),
  methods: {
    async showNewPhotos() {
      this.loading = true
      this.page++
      this.newPhotos = await this.$store.dispatch('searchPhotos', {
        tagName: this.tagName,
        page: this.page
      })
      this.newPhotos.map(newPhoto => this.photosList.push(newPhoto))
      this.loading = false
    },
    async searchByOrientation() {
      await this.$store.dispatch('searchPhotos', {
        tagName: this.tagName,
        page: this.page,
        orientation: this.selectOrientation
      })
      bus.$emit('change-photos', this.photosListByOrientation)
    }
  },
  computed: {
    tagName() {
      return this.$store.getters.tagName
    },
    photosListByOrientation() {
      return this.$store.getters.photos
    }
  }
}
</script>
