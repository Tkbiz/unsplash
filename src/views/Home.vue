<template>
  <div>
    <Loader v-if="loading"/>
    <div v-else>
      <MainBanner
          :tags="tags"
          class="pb-md-16"/>
      <section
          v-if="showErrorMessage"
          class="container error-block">
        <v-row justify="center">
          <h2 class="error-message">
            {{ errorMessage }}
          </h2>
        </v-row>
      </section>
      <Cards
          :userName="logInUser.username"
          :photosList="photosList"/>
    </div>
  </div>
</template>

<script>
import MainBanner from '@/components/MainBanner'
import Cards from '@/components/cards/Cards'
import {bus} from '@/bus'

export default {
  name: 'Home',
  components: {
    MainBanner,
    Cards
  },
  data: () => ({
    showErrorMessage: false,
    photos: [],
    tags: [
      'nature',
      'travel',
      'animals',
      'food',
      'health',
      'technology',
      'events'
    ],
    page: 1,
    loading: true,
    errorMessage: 'Please fill in the correct tag name!'
  }),
  methods: {
    setNewPhotos() {
      bus.$on('change-photos', data => {
        !data.length ? this.showErrorMessage = true : this.showErrorMessage = false
        this.photosList = data
      })
    }
  },
  computed: {
    photosList: {
      get: function () {
        return this.photos
      },
      set: function (newPhotos) {
        this.photos = newPhotos
      }
    },
    accessToken() {
      return this.$store.getters.accessToken
    },
    logInUser() {
      return this.$store.getters.logInUser
    }
  },
  async mounted() {
    const tagsLength = this.tags.length
    let tagName = this.tags[Math.floor(Math.random() * tagsLength)]
    this.photos = await this.$store.dispatch('searchPhotos', {
      tagName: tagName,
      page: this.page
    })
    this.loading = false
    this.$store.commit('fetchTagName', tagName)
    this.setNewPhotos()
  }
}
</script>

<style scoped lang="scss">
.error-block {
  .error-message {
    color: #100e0b;
    font: 36px 'helvetica';
    letter-spacing: 0.15px;
  }
}
</style>
