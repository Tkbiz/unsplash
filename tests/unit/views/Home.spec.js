import {mount, createLocalVue} from '@vue/test-utils'
import Vue from 'vue'
import Vuex from 'vuex'
import Vuetify from 'vuetify'
import Home from '@/views/Home'
import Loader from '@/components/app/Loader'
import MainBanner from '@/components/MainBanner'
import Cards from '@/components/cards/Cards'
import {bus} from '@/bus'

Vue.use(Vuetify)
Vue.component('Loader', Loader)

const localVue = createLocalVue()
localVue.use(Vuex)

describe('Home.vue', () => {
  let vuetify
  let wrapper
  let actions
  let mutations
  let getters
  let store
  const spy = {}

  const observe = jest.fn()
  const unobserve = jest.fn()
  const disconnect = jest.fn()

  window.IntersectionObserver = jest.fn(() => ({
    observe,
    unobserve,
    disconnect
  }))

  beforeEach(() => {
    vuetify = new Vuetify()

    spy.setNewPhotos = jest.spyOn(Home.methods, 'setNewPhotos')

    getters = {
      photosList: () => ([]),
      accessToken: () => false,
      logInUser: () => ({})
    }
    actions = {
      searchPhotos: jest.fn(),
      fetchRandomPhoto: jest.fn()
    }
    mutations = {
      fetchTagName: jest.fn()
    }
    store = new Vuex.Store({
      getters,
      actions,
      mutations
    })

    wrapper = mount(Home, {
      localVue,
      vuetify,
      store,
      data: () => ({
        showErrorMessage: false,
        photos: [],
        tags: [
          'nature',
          'travel',
          'animals',
          'food',
          'health',
          'technology',
          'events'
        ],
        page: 1,
        loading: true,
        errorMessage: 'Please fill in the correct tag name!'
      })
    })
  })

  afterEach(() => {
    wrapper.destroy()
  })

  it('is a Vue instance', () => {
    expect(wrapper.vm).toBeTruthy()
  })

  it('renders MainBanner', () => {
    expect(wrapper.exists(MainBanner)).toBe(true)
  })

  it('renders Loader', () => {
    expect(wrapper.exists(Loader)).toBe(true)
  })

  it('renders Cards', () => {
    expect(wrapper.exists(Cards)).toBe(true)
  })

  it('renders bus event', () => {
    expect(wrapper.exists(bus)).toBe(true)
  })

  it('should match snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders photosList using computed options', () => {
    wrapper.setData({photosList: []})
    expect(JSON.stringify(wrapper.vm.photosList)).toBe(JSON.stringify([]))
  })

  it('dispatches a searchPhotos action on mounted hook', () => {
    expect(actions.searchPhotos).toHaveBeenCalled()
    expect(actions.searchPhotos.mock.calls).toHaveLength(1)
  })

  it('commits tagName on mounted hook', () => {
    expect(mutations.fetchTagName).toHaveBeenCalled()
  })

  it('renders logInUser using computed options', () => {
    expect(JSON.stringify(wrapper.vm.logInUser)).toBe(JSON.stringify({}))
  })

  it('renders accessToken using computed options', () => {
    expect(wrapper.vm.accessToken).toBe(false)
  })

  it('calls setNewPhotos method on mounted hook', () => {
    expect(spy.setNewPhotos).toHaveBeenCalled()
  })
})
