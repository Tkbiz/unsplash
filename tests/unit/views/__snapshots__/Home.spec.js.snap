// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home.vue should match snapshot 1`] = `
<div>
  <div>
    <div class="container--fluid main-banner pb-md-16">
      <div class="row justify-end align-center position-relative ma-0">
        <div class="container banner-block">
          <div class="row">
            <div class="pa-xl-0 col-md-8 col-lg-5 col-xl-4 col-12">
              <!---->
              <!----> <button type="button" class="banner-btn v-btn v-btn--depressed theme--light v-size--default"><span class="v-btn__content">
            explore all
          </span></button>
            </div>
          </div>
        </div>
        <div class="position-relative banner-img col-lg-7 col-12">
          <!---->
        </div>
        <div class="container search-position pa-xl-0">
          <div class="search-nav col col-12">
            <div class="row">
              <div class="v-input custom-input v-input--hide-details theme--light v-text-field">
                <div class="v-input__control">
                  <div class="v-input__slot">
                    <div class="v-text-field__slot"><label for="input-149" class="v-label theme--light" style="left: 0px; position: absolute;">Search for high-resolution photos</label><input id="input-149" type="text"></div>
                    <div class="v-input__append-inner">
                      <div class="v-image v-responsive search-icon theme--light">
                        <transition-stub name="fade-transition" mode="in-out">
                          <div class="v-image__image v-image__image--preload v-image__image--cover" style="background-position: center center;"></div>
                        </transition-stub>
                        <div class="v-responsive__content"></div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div class="trending">
                Trending
              </div>
              <div class="v-item-group theme--light v-slide-group v-chip-group v-chip-group--column">
                <div class="v-slide-group__prev v-slide-group__prev--disabled">
                  <transition-stub name="fade-transition"></transition-stub>
                </div>
                <div class="v-slide-group__wrapper">
                  <div class="v-slide-group__content"><span tabindex="0" class="v-chip v-chip--clickable v-chip--no-color theme--light v-size--default"><span class="v-chip__content">
          nature
        </span></span><span tabindex="0" class="v-chip v-chip--clickable v-chip--no-color theme--light v-size--default"><span class="v-chip__content">
          travel
        </span></span><span tabindex="0" class="v-chip v-chip--clickable v-chip--no-color theme--light v-size--default"><span class="v-chip__content">
          animals
        </span></span><span tabindex="0" class="v-chip v-chip--clickable v-chip--no-color theme--light v-size--default"><span class="v-chip__content">
          food
        </span></span><span tabindex="0" class="v-chip v-chip--clickable v-chip--no-color theme--light v-size--default"><span class="v-chip__content">
          health
        </span></span><span tabindex="0" class="v-chip v-chip--clickable v-chip--no-color theme--light v-size--default"><span class="v-chip__content">
          technology
        </span></span><span tabindex="0" class="v-chip v-chip--clickable v-chip--no-color theme--light v-size--default"><span class="v-chip__content">
          events
        </span></span></div>
                </div>
                <div class="v-slide-group__next v-slide-group__next--disabled">
                  <transition-stub name="fade-transition"></transition-stub>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <!---->
    <div class="container px-xl-0">
      <div class="row mx-xl-n4">
        <div class="d-flex px-xl-4 py-0 mb-8 col-sm-6 col-md-3 col-lg-3 col-12">
          <div class="v-input custom-select theme--light v-text-field v-text-field--enclosed v-text-field--outlined v-select">
            <div class="v-input__control">
              <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-163" class="v-input__slot">
                <fieldset aria-hidden="true">
                  <legend style="width: 0px;"><span>â€‹</span></legend>
                </fieldset>
                <div class="v-select__slot"><label for="input-163" class="v-label theme--light" style="left: 0px; position: absolute;">Any orientation</label>
                  <div class="v-select__selections"><input id="input-163" readonly="readonly" type="text" aria-readonly="false" autocomplete="off"></div>
                  <div class="v-input__append-inner">
                    <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
                  </div><input type="hidden">
                </div>
                <div class="v-menu">
                  <!---->
                </div>
              </div>
              <div class="v-text-field__details">
                <div class="v-messages theme--light">
                  <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div class="row mx-xl-n4">
        <!---->
        <div class="observer"></div>
      </div>
    </div>
  </div>
</div>
`;
